packages/field-base/src/clear-button-mixin.js

```diff
@@ -0,0 +1,78 @@
+/**
+ * @license
+ * Copyright (c) 2021 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import { dedupingMixin } from '@polymer/polymer/lib/utils/mixin.js';
+import { InputMixin } from './input-mixin.js';
+
+const ClearButtonMixinImplementation = (superclass) =>
+  class ClearButtonMixinClass extends InputMixin(superclass) {
+    static get properties() {
+      return {
+        /**
+         * Set to true to display the clear icon which clears the input.
+         * @attr {boolean} clear-button-visible
+         */
+        clearButtonVisible: {
+          type: Boolean,
+          reflectToAttribute: true,
+          value: false
+        }
+      };
+    }
+
+    /** @protected */
+    get _clearOnEsc() {
+      return true;
+    }
+
+    /** @protected */
+    ready() {
+      super.ready();
+
+      this.addEventListener('keydown', (e) => this._handleKeyDown(e));
+
+      this._clearElement = this.$.clearButton;
+      this._clearElement.addEventListener('click', this._onClearButtonClick.bind(this));
+    }
+
+    /**
+     * Clear the value of this field.
+     */
+    clear() {
+      this.value = this._inputNode.value = '';
+    }
+
+    /** @private */
+    _onClearButtonClick(e) {
+      e.preventDefault();
+      this._inputNode.focus();
+      this.clear();
+      const inputEvent = new Event('input', { bubbles: true, composed: true });
+      inputEvent.__fromClearButton = true;
+
+      const changeEvent = new Event('change', { bubbles: true });
+      changeEvent.__fromClearButton = true;
+
+      this._inputNode.dispatchEvent(inputEvent);
+      this._inputNode.dispatchEvent(changeEvent);
+    }
+
+    /**
+     * @param {Event} event
+     * @protected
+     */
+    _handleKeyDown(event) {
```

Comment:
This would be a very common name for a keyboard event handler in any component, which means this might easily be overriden by a custom component using this mixin. Especially if we want the mixins to be reusable by other parties. There are lots of other method names in the new mixins (not only from this PR) that could clash with other mixins that have similar functionality.

Why not build the mixins in a way that these method names and properties are guaranteed to be unique? I know there is some way to do it with Symbols, there might be other solutions.

For example, see the approach taken here: 
- Symbols: https://github.com/elix/elix/blob/main/src/base/internal.js
- Usage: https://github.com/elix/elix/blob/main/src/base/FocusCaptureMixin.js

Also if we ever introduce a method like this into InputMixin, or use this mixin on top of another one that has this method, then we would overwrite that, because there is no super call here. No test would catch that, because InputMixin only tests itself, and ClearButtonMixin doesn't test that InputMixin key down still works. If we want these to be composable we need to think about every method possibly being overriden or overriding something.

Reply:
Now I'm thinking that giving such a specific name to the listener may be a real problem when using this mixin with other components or mixins.

Basic search of the `_onKeyDown()` method across `*.js` files returns ~14 definitions. 

At least, should `ClearButtonMixin._onKeyDown()` call `super`? It doesn't currently do that.

Anyway, we need to make sure that no other components and mixins define their own `_onKeyDown()` without calling `super`.

Reply:
Another problem with using the same listener name across different mixins is that the listener might potentially be attached to an element multiple times. Especially if using `bind` at the moment of adding the event listener to an element. 

Consider this pseudo-code example:

```js
class FirstMixin extends PolymerElement {
  connectedCallback() {
    super.connectedCallback();
    
    this.addEventListener(
      'keydown', 
      this._onKeyDown.bind(this) /* <--- creates a copy of the method with bound `this` */
    );
  }
  
  _onKeyDown() {
    super?._onKeyDown();
    
    ...
  }
}

class SecondMixin extends FirstMixin, PolymerElement {
  connectedCallback() {
    super.connectedCallback();
    
    this.addEventListener(
      'keydown', 
      this._onKeyDown.bind(this) /* <--- creates another copy of the method with bound `this` */
    );
  }
  
  _onKeyDown() {
    super?._onKeyDown();
    
    ...
  }
}
```

The `_onKeyDown` method, in this case, will be called twice once a key is pressed.

Most likely it is a rare case, though we still need to be careful.

Reply:
I think we could add `super._onKeyDown` call in all the other places where this mixin is used. 
Alternatively, we could create `KeyboardMixin` as in the [prototype](https://github.com/vaadin/component-mixins/blob/master/packages/keyboard-mixin/keyboard-mixin.ts).

Let's create another issue for it so we could this proceed with this PR.
Note, I have updated a clear button listener to not use `.bind(this)`.

Reply:
We could avoid the mentioned problem by binding methods in the constructor.

```js
constructor() {
  this._onKeyDown = this._onKeyDown.bind(this);
}

ready() {
  this.addEventListener('keydown', this._onKeyDown);
}
```

or

```js
constructor() {
  this._boundOnKeyDown = this._onKeyDown.bind(this);
}

ready() {
  this.addEventListener('keydown', this._boundOnKeyDown);
}
```

In this case, the listener will be attached exactly once even if another mixin would try to add the listener once again.